src/Application/Kernel.php

```diff
@@ -161,4 +149,57 @@ private function initDiscovery(Container $container): void
             $discovery->storeCache();
         }
     }
+
+    private function discoverInstalledPackageLocations(): void
+    {
+        $composerPath = path($this->root, 'vendor/composer');
+        $installed = path($composerPath, 'installed.json');
+
+        if (is_file($installed)) {
+            $installed = $this->loadJsonFile($installed);
+
+            foreach ($installed['packages'] ?? [] as $package) {
+                $packagePath = realpath(path($composerPath, $package['install-path']));
+
+                if (isset($package['require']['tempest/framework'], $package['autoload']['psr-4'])
+                    && is_array($package['autoload']['psr-4'])
+                ) {
+                    foreach ($package['autoload']['psr-4'] as $namespace => $namespacePath) {
+                        $namespacePath = path($packagePath, $namespacePath);
+                        $this->addDiscoveryLocation($namespace, $namespacePath);
+                    }
+                }
+            }
+        }
+    }
+
+    private function discoverTempestNamespaces(): void
+    {
+        $composer = $this->loadJsonFile(path($this->root, 'composer.json'));
+
+        $autoloads = [$composer['autoload']['psr-4'] ?? []];
+        if (env('ENVIRONMENT') === 'production') {
+            $autoloads[] = $composer['autoload']['psr-4'] ?? [];
+        }
+        $autoload = array_merge(...$autoloads);
+
+        foreach ($autoload as $namespace => $path) {
+            $path = path($this->root, $path);
+            $this->addDiscoveryLocation($namespace, $path);
+        }
+    }
+
+    private function addDiscoveryLocation(string $namespace, string $path): void
+    {
+        $this->appConfig->discoveryLocations[] = new DiscoveryLocation(
+            namespace: $namespace,
+            path     : $path,
+        );
+    }
+
+    private function loadJsonFile(string $path): array
+    {
+        return json_decode(file_get_contents($path), true)
```

Comment:
Does this need a separate function?

Reply:
i took your suggestion for a `KernelException` class to throw when the target file isn't found, and put it here. It cleans this up a little _and_ the spot where `is_file()` was being checked. One less nesting level above, and more consistent flow here: always check for the file, then return the decoded file contents.

Does that make more sense?